### PR TITLE
Unmanaged dynamic objects

### LINKED ISSUE #708: Import and update object graphs with multiple identical primary keys
The cache maintained during import of objects doesn't recognize that multple objects can have the same primary key and theoretically should update each other. Added a test for this in #696 as a replica of an realm-java issue https://github.com/realm/realm-java/issues/4957

### PR DESCRIPTION
**Pull request by rorbech**
This adds the ability to create unmanaged dynamic objects with 
```
public fun DynamicRealmObject(type: String, vararg properties: Pair<String, Any?>)
```
And replaces old `DynamicMutableRealm.create` with  
```
public interface DynamicMutableRealm : DynamicRealm {
    public fun copyToRealm(obj: BaseRealmObject): DynamicMutableRealmObject
}
```

TODOs:
- [x] Should a missing primary key property be interpreted as null or throw?
- [x] Import of lists
- [x] Initialize schema metadata with companion property references when caching schema details

Closes #708 and #808

### CODE REVIEW

File: test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmTests.kt [hunk 1]
```diff
@@ -79,60 +88,198 @@ class DynamicMutableRealmTests {
...
+    //     mapOf(
+    //         "name" to "HELLO",
+    //         "child" to dynamicMutableObject.create("Sample")
+    //     )
+    // )
+
     // TODO Add variants for each type
     @Test
-    fun createPrimaryKey() {
+    fun copyToRealm_withPrimaryKey() {
```
Comment (cmelchior) [inline test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmTests.kt:174]: We would need to test for a lot of error cases, like unknown properties, wrong types etc.

**Review by cmelchior — COMMENTED** [review]
Just left some smaller comments, but generally, this looks good.

We do need to settle on whether to use constructor functions or not, or at least decide in which cases (or not) we want them. Maybe have them for all data types like RealmInstant, ObjectId, UUID, but not for other API classes (Hello `Realm.open`).

But we should probably discuss it in person.

File: test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmTests.kt (hunk 1 above)
Comment (rorbech) [reply test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmTests.kt:102]: There are already cases to check that verifies that we throw when object type and property name does exist and if property type and primary key is of wrong type!?

File: packages/library-base/src/commonMain/kotlin/io/realm/internal/dynamic/DynamicUnmanagedRealmObject.kt [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import io.realm.BaseRealmObject
+import io.realm.RealmList
+import io.realm.dynamic.DynamicMutableRealmObject
+import io.realm.internal.RealmObjectInternal
+import io.realm.internal.RealmObjectReference
+import io.realm.realmListOf
+import kotlin.reflect.KClass
+
+internal class DynamicUnmanagedRealmObject(override val type: String,
+    // FIXME Should we expose this as mutable? And then drop all the other getters/setters?
```
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/internal/dynamic/DynamicUnmanagedRealmObject.kt]: Yeah. Old comment. This whole object is actually internal, so things should be accessed/updated through the public API.

File: packages/library-base/src/commonMain/kotlin/io/realm/dynamic/DynamicRealmObject.kt [hunk 3]
```diff
@@ -144,6 +146,13 @@ public interface DynamicRealmObject : BaseRealmObject {
     public fun getObjectList(propertyName: String): RealmList<out DynamicRealmObject>
 }
 
+public fun DynamicRealmObject(type: String, vararg properties: Pair<String, Any?>): DynamicMutableRealmObject {
```
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/dynamic/DynamicRealmObject.kt]: According to discussions I added `DynamicMutableRealmObject.create` instead of this. We can tweak if we find the above pattern better in the future, but for now just avoiding it to ensure that we don't have diverging patterns. If this is really annoying to write people can just do the extension themselves.

File: test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmTests.kt [hunk 4]
```diff
@@ -79,60 +85,192 @@ class DynamicMutableRealmTests {
...
+        )
+        dynamicMutableRealm.query("Sample", "stringField = 'PARENT'").find().single().run {
+            getObject("nullableObject")!!.run {
+                assertEquals("CHILD", getValue("stringField"))
+            }
+        }
+    }
+
+    @Test
+    fun copyToRealm_withUnmanagedTypedObject() {
```
Comment (cmelchior) [inline test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmTests.kt:160]: You mean dynamic object?

**Review by cmelchior — COMMENTED** [review]
I tried to review the PR, but I must confess that a lot of the implementation details were super hard to follow, so I relied mostly on tests that seems solid.

Missing changelog update.

File: test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmObjectTests.kt [hunk 5]
```diff
@@ -91,10 +92,24 @@ class DynamicMutableRealmObjectTests {
         PlatformUtils.deleteTempDir(tmpDir)
     }
 
+    @Test
+    fun get_returnsDynamicMutableObject() {
```
[the remark this answers is not in the document]
Comment (rorbech) [reply test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmObjectTests.kt:96]: Yep: https://github.com/realm/realm-kotlin/blob/master/test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmObjectTests.kt#L96

File: test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmTests.kt (hunk 4 above)
Comment (rorbech) [reply test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmTests.kt:160]: No. We can also import typed objects, so added a test for that 🤷 ... is it too confusing that we allow that? 🤔

File: test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmTests.kt [hunk 6]
```diff
@@ -79,60 +85,192 @@ class DynamicMutableRealmTests {
...
+    // @Test
+    // fun create_throwsWithPrimaryKey() {
+    //     assertFailsWithMessage<IllegalArgumentException>("Class does not have a primary key)") {
+    //         dynamicMutableRealm.createObject("Sample", "PRIMARY_KEY")
+    //     }
+    // }
+
     @Test
-    fun createPrimaryKey_throwsWithWrongPrimaryKeyType() {
+    // FIXME Should we allowing interpreting unset primary key property as null or should we throw?
```
[the remark this answers is not in the document]
Comment (rorbech) [reply test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmTests.kt]: Added a guard for it so that it requires an explicit primary key property even for `null`.

File: packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt [hunk 7]
```diff
@@ -68,33 +76,42 @@ internal object RealmObjectHelper {
...
+            realmObjectCompanion != null,
+        )
+    }
 
     // Cannot call managedRealmList directly from an inline function
     internal fun <R> getListByKey(
         obj: RealmObjectReference<out BaseRealmObject>,
         key: io.realm.internal.interop.PropertyKey,
         elementType: KClass<*>,
+        isObjectList: Boolean,
```
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt:110]: It is a bit difficult to share as much code as possible without this dynamic path. The list operator requires the native pointer amongst others to avoid passing those around internally on all the list operations and this makes it a bit annoying to create the operator completely separate from the managed list. I have tried to extract building the operator into a separate method and inline some of the calls. Hopefully it helps. ... until we also need to take embeddedness into account 🤪

File: packages/library-base/src/commonMain/kotlin/io/realm/dynamic/DynamicMutableRealmObject.kt [hunk 8]
```diff
@@ -45,4 +46,20 @@ public interface DynamicMutableRealmObject : DynamicRealmObject {
      * name, or if the value doesn't match the [RealmStorageType.kClass] type of the property.
      */
     public fun <T> set(propertyName: String, value: T): DynamicMutableRealmObject
+    public fun set(vararg pairs: Pair<String, Any?>): DynamicMutableRealmObject {
```
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/dynamic/DynamicMutableRealmObject.kt:49]: Just like for `DynamicMutableRealmObject.create` it is just convenience for setting multiple properties in one go
```
object.set(
    "id" to 1,
    "name" to "NAME,
)
```
instead of 
```
object.apply {
    set("id", 1)
    set("name", "NAME)
}
```

File: test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmTests.kt (hunk 4 above)
Comment (nhachicha) [reply test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmTests.kt:160]: If we allow this is there a test that makes sure they're using the same version?

**Review by nhachicha — COMMENTED** [review]
Massive PR 👍 I tried to look for obvious things, but it looks good besides some minor stuff. Nice effort 🥇

File: packages/library-base/src/commonMain/kotlin/io/realm/dynamic/DynamicRealmObject.kt [hunk 9]
```diff
@@ -17,6 +17,7 @@
 package io.realm.dynamic
 
 import io.realm.BaseRealmObject
+import io.realm.Realm
```
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/dynamic/DynamicRealmObject.kt:20]: No, it is referenced in the dokka.

File: test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmTests.kt (hunk 4 above)
Comment (rorbech) [reply test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmTests.kt:160]: So, you would like to enforce that `schemaVersion` is the same for both? I don't think we have a way to enforce that, since there is no schema version tied to the unmanaged object. The schema of that is exactly whats in the model definition. I guess we could just iterate the class metadata for it but it sounds like a bit effort. If people do it I would assume that they know what they do ... maybe better to block this completely ... but it is just awfully convenient to be able to just create objects from the typed one in a dynamic realm!!??

File: test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmObjectTests.kt [hunk 10]
```diff
@@ -91,10 +94,26 @@ class DynamicMutableRealmObjectTests {
...
+                "Sample",
+                "stringField" to "PARENT",
+                "nullableObject" to DynamicMutableRealmObject.create("Sample", "stringField" to "CHILD")
+            )
+        )
+        assertTrue(parent.isManaged())
+        val child: DynamicMutableRealmObject? = parent.getObject("nullableObject")
+        assertNotNull(child)
+        assertTrue(child.isManaged())
+        child.set("stringField", "UPDATED_CHILD")
```
[the remark this answers is not in the document]
Comment (rorbech) [reply test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmObjectTests.kt:110]: That `child` actually has the contents of the imported object? There are plenty of tests around that objects are imported correctly. This just verifies that objects retrieved from a mutable object are mutable objects themselves.

File: packages/library-base/src/commonMain/kotlin/io/realm/internal/Converters.kt [hunk 11]
```diff
@@ -190,24 +191,32 @@ internal inline fun <T : BaseRealmObject> realmValueToRealmObject(
...
+    return RealmValue(
+        value?.let {
+            val realmObjectReference = value.realmObjectReference
+            // If managed ...
+            if (realmObjectReference != null) {
+                // and from the same version we just use object as is
+                if (realmObjectReference.owner == realmReference) {
+                    value
+                } else {
+                    throw IllegalArgumentException("asdf")
```
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/internal/Converters.kt]: Ups 🙈 ... As per discussion we could just copy the values, but I think this is not necessarily intentional from user perspective. If they fail to realize that it is an old version and it gets copied they could end up with multiple similar objects. I will leave it to throw (with a more appropriate message 🤪) then we can discuss whether we should offer options to automatically resolve objects to their latest version, etc. later.

[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/internal/Converters.kt]: Also reenabled the test for this in https://github.com/realm/realm-kotlin/pull/832/files#diff-10fb347be8a56caadcfa75369423ecbd0674592a074ec57de6300e9caee36873R183

File: packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt [hunk 12]
```diff
@@ -168,25 +195,139 @@ internal object RealmObjectHelper {
...
+    internal fun assignDynamic(
+        target: DynamicMutableRealmObject,
+        source: BaseRealmObject,
+        updatePolicy: MutableRealm.UpdatePolicy,
+        cache: ObjectCache
+    ) {
+        val properties: List<Pair<String, Any?>> = if (source is DynamicUnmanagedRealmObject) {
+            source.properties.toList()
+        } else if (source is DynamicRealmObject) {
+            TODO("Cannot import managed dynamic objects")
```
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt]: We can actually not end up here. It was left over because I was imagining that I had to handle the case as part of importing dynamic embedded objects, but we should basically never end here. Added an explanation of why.

File: test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmObjectTests.kt [hunk 13]
```diff
@@ -301,30 +320,45 @@ class DynamicMutableRealmObjectTests {
...
-        assertNotNull(child)
-        child.set("stringField", "UPDATED_CHILD")
+    fun set_detectsDuplicates() {
+        val child = DynamicMutableRealmObject.create(
+            "Sample",
+            "stringField" to "child"
+        )
+        val intermediate = DynamicMutableRealmObject.create(
+            "Sample",
+            "stringField" to "intermedidate",
```
[the remark this answers is not in the document]
Comment (rorbech) [reply test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmObjectTests.kt:353]: Added https://github.com/realm/realm-kotlin/pull/832/files#diff-b1a7ff639ec13d0894ccca75fb59c1bd6be2eb5da205018b533a42d8e37dd0a8R114 to ensure it is covered.

File: test/base/src/androidTest/kotlin/io/realm/test/shared/ImportTests.kt [hunk 14]
```diff
@@ -199,12 +198,10 @@ class ImportTests {
         }
 
         val importedRoot = realm.writeBlocking {
-            copyToRealm(unmanaged)
+            assertFailsWithMessage<IllegalArgumentException>("Cannot import an outdated object") {
+                copyToRealm(unmanaged)
```
Comment (cmelchior) [inline test/base/src/androidTest/kotlin/io/realm/test/shared/ImportTests.kt:202]: Having the variable named `unmanaged` here is a bit weird and is messing with my head.. It is correct, but maybe calling it something like `outdatedContainer` or `unmanagedWithOldManagedObj` would make it a little more clear.

File: test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmTests.kt (hunk 4 above)
Comment (cmelchior) [reply test/base/src/androidTest/kotlin/io/realm/test/shared/dynamic/DynamicMutableRealmTests.kt:160]: IMO the current behavior seems fine. Dynamic objects are by their nature a little more brittle and depend on people knowing what they are doing

**Review by cmelchior — APPROVED** [review]
Nice...a lot of great improvements here. Only had some last minor comments. Otherwise :shipit:

File: packages/library-base/src/commonMain/kotlin/io/realm/dynamic/DynamicMutableRealmObject.kt [hunk 15]
```diff
@@ -45,4 +46,39 @@ public interface DynamicMutableRealmObject : DynamicRealmObject {
...
+    public companion object {
+        /**
+         * Create an unmanaged dynamic object.
+         *
+         * The type and properties are only checked when the object is imported through [DynamicMutableRealm.copyToRealm].
+         *
+         * @param type the class name of the object.
+         * @param properties properties of the object.
+         *
+         * @see DynamicMutableRealm.copyToRealm
```
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/dynamic/DynamicMutableRealmObject.kt:64]: Yes, just like for Javadoc you don't need more to link to reference. It is also navigatable in the IDE.

File: packages/library-base/src/commonMain/kotlin/io/realm/dynamic/DynamicMutableRealm.kt [hunk 16]
```diff
@@ -19,6 +19,7 @@ package io.realm.dynamic
 import io.realm.BaseRealmObject
 import io.realm.Deleteable
 import io.realm.MutableRealm
+import io.realm.MutableRealm.UpdatePolicy
```
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/dynamic/DynamicMutableRealm.kt]: Yep. I moved it to the top level.
